File: src/pythonic/balance_multiply.py
class BalanceMultiply:
    def balancePair(self, numA, numB):
        accumulator = 0

        while True:
            if numA < numB:
                numA, numB = numB, numA

            if numA == 0 or numB == 0:
                return (numA, numB, accumulator)

            if numA % 2 == 1 and numB % 2 == 1:
                if numA // numB > 1:
                    accumulator += (numA - 1) + numB
                    numA = (numA - 1) >> 1
                    numB = (numB - 1) << 1
                    continue

            if numA % 2 == 0 and numB % 2 == 0:
                if numA // numB > 1:
                    numA = numA >> 1
                    numB = numB << 1
                    continue

            if numA % 2 == 1 and numB % 2 == 0:
                if numA // numB > 1:
                    accumulator += numB
                    numA = (numA - 1) >> 1
                    numB = numB << 1
                    continue

            if numA // numB > 1:
                accumulator += numA
                numA = numA >> 1
                numB = (numB - 1) << 1
                continue
            return (numA, numB, accumulator)

    def square(self, num):
        pow = 0
        res = 0
        tmp = num
        while (tmp):
            if (tmp & 1):
                res += (num << pow)
            pow += 1
            tmp = tmp >> 1
        return res

    def solve(self, numA, numB):
        if numA == 0 or numB == 0:
            return 0
        result = 0
        if numA < numB:
            numA, numB = numB, numA

        while numB > 0:
            quotient = numA // numB
            result += quotient * self.square(numB)
            numA, numB = numB, numA % numB
        return result

    def multiply(self, numA, numB):
        ans = self.balancePair(numA, numB)
        return self.solve(ans[0], ans[1]) + ans[2]

File: src/balance_multiply.rs
extern crate num_bigint;
extern crate num_traits;

use num_bigint::BigUint;
use num_traits::{One, Zero};

pub struct BalanceMultiply;

impl BalanceMultiply {
    pub fn balance_pair(
        &self,
        mut num_a: BigUint,
        mut num_b: BigUint,
    ) -> (BigUint, BigUint, BigUint) {
        let mut accumulator = BigUint::zero();

        while !num_a.is_zero() && !num_b.is_zero() {
            if num_a < num_b {
                std::mem::swap(&mut num_a, &mut num_b);
            }

            if num_a.is_zero() || num_b.is_zero() {
                return (num_a, num_b, accumulator);
            }

            let num_a_is_odd = &num_a & BigUint::one() != BigUint::zero();
            let num_b_is_odd = &num_b & BigUint::one() != BigUint::zero();

            if num_a_is_odd && num_b_is_odd {
                if &num_a / &num_b > BigUint::one() {
                    accumulator += (&num_a - BigUint::one()) + &num_b;
                    num_a = (&num_a - BigUint::one()) >> 1;
                    num_b = (&num_b - BigUint::one()) << 1;
                    continue;
                }
            }

            if !num_a_is_odd && !num_b_is_odd {
                if &num_a / &num_b > BigUint::one() {
                    num_a >>= 1;
                    num_b <<= 1;
                    continue;
                }
            }

            if num_a_is_odd && !num_b_is_odd {
                if &num_a / &num_b > BigUint::one() {
                    accumulator += &num_b;
                    num_a = (&num_a - BigUint::one()) >> 1;
                    num_b <<= 1;
                    continue;
                }
            }

            if &num_a / &num_b > BigUint::one() {
                accumulator += &num_a;
                num_a >>= 1;
                num_b = (&num_b - BigUint::one()) << 1;
                continue;
            }

            break;
        }

        (num_a, num_b, accumulator)
    }

    pub fn square(&self, num: &BigUint) -> BigUint {
        let mut pow = 0u64;
        let mut res = BigUint::zero();
        let mut tmp = num.clone();

        while !tmp.is_zero() {
            if &tmp & BigUint::one() != BigUint::zero() {
                res += num << pow;
            }
            pow += 1;
            tmp >>= 1;
        }

        res
    }

    pub fn solve(&self, mut num_a: BigUint, mut num_b: BigUint) -> BigUint {
        if num_a.is_zero() || num_b.is_zero() {
            return BigUint::zero();
        }

        let mut result = BigUint::zero();

        if num_a < num_b {
            std::mem::swap(&mut num_a, &mut num_b);
        }

        while num_b > BigUint::zero() {
            let quotient = &num_a / &num_b;
            result += &quotient * self.square(&num_b);
            let remainder = &num_a % &num_b;
            num_a = num_b;
            num_b = remainder;
        }

        result
    }

    pub fn multiply(&self, num_a: BigUint, num_b: BigUint) -> BigUint {
        let (balanced_a, balanced_b, accumulator) = self.balance_pair(num_a, num_b);
        self.solve(balanced_a, balanced_b) + accumulator
    }
}
